### PR TITLE
feat(dapi-client): handle asynchronous errors

### PR DESCRIPTION
**Pull request by markin-io**
## Issue being fixed or feature implemented

The feature improves error handling for DAPI-client by making it an EventEmitter to handle asynchronous errors.
A few fixes also applied to properly dispatch error events from within the BlockHeadersProvider.

## What was done?
- Refactor reading of historical data in BlockHeadersReader
- Write tests to cover error handling
- Refactor tests
- Make DAPIClient an EventEmitter
- Bubble ERROR events from within BlockHeadersProvider to the DAPIClient

## How Has This Been Tested?

Added relevant unit and integration tests

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersProvider.js [hunk 1]
```diff
@@ -19,17 +20,23 @@ const defaultOptions = {
   autoStart: false,
 };
 
-class BlockHeadersProvider {
+const EVENTS = {
+  ERROR: 'ERROR',
```
Comment (shumkov) [inline packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersProvider.js:24]: I think we should use `error` name of the event to enable default error handling logic https://nodejs.org/api/events.html#error-events

File: packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js [hunk 2]
```diff
@@ -33,6 +47,14 @@ class BlockHeadersReader extends EventEmitter {
    * @returns {Promise<void>}
    */
   async readHistorical(fromBlockHeight, toBlockHeight) {
+    if (this.historicalStreams.length) {
+      this.removeAllListeners(COMMANDS.HANDLE_STREAM_RETRY);
```
Comment (shumkov) [inline packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js:51]: Maybe we should just throw an error here so you need manually reset first and then call `readHistorical` again?

File: packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js [hunk 3]
```diff
@@ -95,13 +140,14 @@ class BlockHeadersReader extends EventEmitter {
   }
 
   /**
-   * A HOC that returns a function to subscribe to block headers and chain locks
+   * A HOF that returns a function to subscribe to historical block headers and chain locks
```
Comment (shumkov) [inline packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js:143]: What is HOF? ))

File: packages/js-dapi-client/lib/DAPIClient.js [hunk 4]
```diff
@@ -45,16 +53,26 @@ class DAPIClient {
     this.core = new CoreMethodsFacade(jsonRpcTransport, grpcTransport);
     this.platform = new PlatformMethodsFacade(grpcTransport);
 
+    this.initBlockHeadersProvider();
+  }
+
```
Comment (shumkov) [inline packages/js-dapi-client/lib/DAPIClient.js:58]: Let's add `@private` JSDoc?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js (hunk 3 above)
Comment (markin-io) [reply packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js:143]: A higher-order function :)

Comment (markin-io) [reply packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js:147]: Googled a bit on the matter. 
Found that there's no convention for naming the HOFs, but only a suggestion to keep it simple. 
Removed "HOF" from the function name after all.

File: packages/js-dapi-client/lib/DAPIClient.js (hunk 4 above)
Comment (markin-io) [reply packages/js-dapi-client/lib/DAPIClient.js:58]: Done

File: packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js (hunk 2 above)
Comment (markin-io) [reply packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersReader.js]: Done

File: packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersProvider.js (hunk 1 above)
Comment (markin-io) [reply packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersProvider.js]: Done

**Comment by markin-io** [comment]
@jawid-h added tests for DAPIClient as you requested

**Review by jawid-h — APPROVED** [review]
Looks legit

**Review by shumkov — APPROVED** [review]
Good job! 👍

**Pull request merged**